File: 2015/Day10/src/day10.py
import re
import sys

DEBUG = True

def solve_puzzle(num_loops, input_string):
    cache = {}
    for i in range(num_loops):
        output_string = ""
        index = 1
        prev_char = input_string[0]
        while(index <= len(input_string)):
            count = 1
            while(index < len(input_string) and input_string[index] == prev_char ):
                index += 1
                count += 1
            result = f"{count}{prev_char}"
            output_string += result
            if(index < len(input_string)):
                prev_char = input_string[index]
            index += 1
            if(DEBUG):
                print(f"{i=}, {index=}, {input_string=}, {output_string=}")
        if(DEBUG):
            print(f"assigning {output_string} to output string")
        input_string = output_string
    return len(input_string)

def solve_puzzle_regex(num_loops ,input_string):
    pattern = r"((.)\2*)"
    cache = {}
    for i in range(num_loops):
        matches = re.findall(pattern, input_string)
        output_string = ""
        for match in matches:
            result = ""
            if match[0] in cache:
                result = cache[match[0]]
            else:
                result = f"{len(match[0])}{match[1]}"
                cache[match[0]] = result
            output_string += result
        input_string = output_string
    if(DEBUG):
        print(f"regex result: {len(input_string)}")
    return len(input_string)

if __name__ == "__main__":
    if(len(sys.argv) != 2):
        print("Usage: python3 day10.py testinput.txt")
        exit(1)
    input_string = ""
    with open(sys.argv[1], "r") as fileobj:
        input_string = fileobj.readline().rstrip()
    if(DEBUG):
        print(f"Input string is {input_string}")
    part1_ans = solve_puzzle_regex(40, input_string)
    print(f"The answer to part 1 is {part1_ans}")
    part2_ans = solve_puzzle_regex(50, input_string)
    print(f"The answer to part 2 is {part2_ans}")

File: 2015/Day10/src/day10.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <regex>
#include <sstream>

#define DEBUG false

using namespace std;

int solve_puzzle(int num_loops, string input_string)
{
    for(int loop_index = 0; loop_index < num_loops; loop_index++)
    {
        ostringstream output_string;
        int count = 1;
        int i;
        char prev_char = input_string.at(0);
        for(i = 1; i < input_string.size(); i++)
        {
            if(input_string.at(i) != prev_char)
            {
                output_string << count << prev_char;
                prev_char = input_string.at(i);
                count = 1;
            }
            else
            {
                count++;
            }
        }
        output_string << count << input_string.at(i - 1);
        input_string = output_string.str();
        if(DEBUG)
        {
            cout << "Loop " << loop_index << ": Output string " << input_string << endl;
        }
    }
    return input_string.size();
}

int main(int argc, char * argv[])
{
    if(argc != 2)
    {
        cout << "Usage: ./day10 input.txt\n";
    }
    ifstream file(argv[1]);
    string line;
    getline(file, line);
    if(DEBUG)
    {
        cout << "The input is " << line << endl;
    }
    file.close();
    if(DEBUG)
    {
        int test_result = solve_puzzle(10, line);
        cout << "The test result is " << test_result << endl;
    }
    int part1_result = solve_puzzle(40, line);
    cout << "The answer to part 1 is " << part1_result << endl;
    int part2_result = solve_puzzle(50, line);
    cout << "The answer to part 2 is " << part2_result << endl;
}
